File: src/DataScience/Julia/example3.jl
abstract type Report end

struct PDFReport <: Report end
struct HTMLReport <: Report end

function generate(r::PDFReport)
    println("Generating PDF report")
end

function generate(r::HTMLReport)
    println("Generating HTML report")
end

abstract type ReportFactory end

struct PDFReportFactory <: ReportFactory end
struct HTMLReportFactory <: ReportFactory end

function create_report(::PDFReportFactory)::Report
    return PDFReport()
end

function create_report(::HTMLReportFactory)::Report
    return HTMLReport()
end

function use_factory(factory::ReportFactory)
    report = create_report(factory)
    generate(report)
end

# Usage
use_factory(PDFReportFactory())
use_factory(HTMLReportFactory())

File: src/Scripting/PythonPY/example3.py
class Report:
    def generate(self):
        pass

class PDFReport(Report):
    def generate(self):
        print("Generating PDF report")

class HTMLReport(Report):
    def generate(self):
        print("Generating HTML report")

class ReportFactory:
    def create_report(self):
        pass

class PDFReportFactory(ReportFactory):
    def create_report(self):
        return PDFReport()

class HTMLReportFactory(ReportFactory):
    def create_report(self):
        return HTMLReport()

def use_factory(factory):
    report = factory.create_report()
    report.generate()

# Usage
use_factory(PDFReportFactory())
use_factory(HTMLReportFactory())
